Library/ViewModels/SurveyResponseViewModel.swift

```diff
@@ -68,33 +83,47 @@ public final class SurveyResponseViewModel: SurveyResponseViewModelType {
       self.closeButtonTappedProperty.signal
     )
 
-    self.goToProject = self.shouldStartLoadProperty.signal.skipNil()
-      .map { request, _ -> (Param, RefTag?)? in
+    self.goToProject = self.policyForNavigationActionProperty.signal.skipNil()
+      .map { action in action.request }
+      .map { request -> (Param, RefTag?)? in
         if case let (.project(param, .root, refTag))? = Navigation.match(request) {
           return (param, refTag)
         }
         return nil
       }
       .skipNil()
 
-    self.shouldStartLoadResponseProperty <~ self.shouldStartLoadProperty.signal.skipNil()
-      .map { request, _ in
-        if !AppEnvironment.current.apiService.isPrepared(request: request) {
+    self.policyDecisionProperty <~ self.policyForNavigationActionProperty.signal.skipNil()
+      .map { action in
+        if !AppEnvironment.current.apiService.isPrepared(request: action.request) {
           return false
         }
 
-        return isSurvey(request: request)
+        return isSurvey(request: action.request)
       }
+      .map { $0 ? .allow : .cancel }
 
     self.showAlert = redirectAfterPostRequest
       .mapConst(Strings.Got_it_your_survey_response_has_been_submitted())
 
     self.title = self.viewDidLoadProperty.signal
       .mapConst(Strings.Survey())
 
+    self.extractFormDataWithJavaScript = surveyResponse
+      .takeWhen(postRequest.filter { $0.httpBody == nil })
+      .map { surveyResponse in
+        "$('#edit_survey_response_\(surveyResponse.id)').serialize()"
```

Comment:
This is the little bit of JavaScript. For a `SurveyResponse` with an `ID` of `12345` this would be:

```javascript
$('#edit_survey_response_12345').serialize()
```
It makes the following assumptions:
1. jQuery is available on the page.
1. There is a form on the page with the ID of `edit_survey_response_<survey_response_id>`.

It returns the serialized form input fields like:

```

survey_response[answers_attributes][1][survey_question_id] | 2610391
survey_response[answers_attributes][1][response] |  
survey_response[answers_attributes][2][survey_question_id] | 2610392
survey_response[answers_attributes][2][response] | small
survey_response[answers_attributes][3][survey_question_id] | 2610393
survey_response[answers_attributes][3][response] |  
survey_response[answers_attributes][4][survey_question_id] | 2610394
survey_response[answers_attributes][4][response][] |  
survey_response[answers_attributes][4][response][] | red
```

Reply:
Note there is an alternative to this `jQuery` snippet that alleviates at least one of these assumptions, but it looks like this 😬 

```swift
self.extractFormDataWithJavaScript = surveyResponse
  .takeWhen(postRequest.filter { $0.httpBody == nil })
  .map { surveyResponse in
    serializeFormDataJavaScript(withSurveyResponseId: surveyResponse.id)
  }

private func serializeFormDataJavaScript(withSurveyResponseId id: Int) -> String {
  // swiftlint:disable line_length
  let serialize = """
  var serialize=function(form){var serialized=[];for(var i=0;i<form.elements.length;i+=1){var field=form.elements[i];
  if(!field.name||field.disabled||field.type==='file'||field.type==='reset'||field.type==='submit'||field.type==='button'){continue}
  if(field.type==='select-multiple'){for(var n=0;n<field.options.length;n+=1){if(!field.options[n].selected){continue}
  serialized.push(encodeURIComponent(field.name)+"="+encodeURIComponent(field.options[n].value))}}
  else if((field.type!=='checkbox'&&field.type!=='radio')||field.checked)
  {serialized.push(encodeURIComponent(field.name)+"="+encodeURIComponent(field.value))}}return serialized.join('&')};
  """
  // swiftlint:enable line_length

  return "\(serialize)\nserialize(document.getElementById('edit_survey_response_\(id)'));"
}
```
This uses a minified `serialize()` function which does the same as jQuery's `serialize()` except we won't be affected by the possible (albeit unlikely) removal of jQuery from the survey response page.